File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int l1 = 0,l2 = 0;
        int k = 0 ;
        int t = 0;
        int n = nums1.length + nums2.length ;
        int prev=-1;
        while(l1 < nums1.length && l2 < nums2.length){
            if( t == n/2 + 1){
                if(n%2 == 0){
                    return (k + prev) / 2.0;
                }
                else return k;
            }
            if(nums1[l1] <= nums2[l2]){
                
                prev=k;
                k = nums1[l1];
                l1 ++;

            }
            else{
                
                prev=k;
                k = nums2[l2];
                l2 ++;

            }t++;
        
        }
        while(l1 < nums1.length) {
            if(t == n/2 + 1)  {if(n%2 == 0){
                    return (k + prev) / 2.0;
                }
                else return k;}
            prev = k;
            k = nums1[l1];
            l1++;
            t++;
        }
         while(l2 < nums2.length) {
            if( t == n/2 + 1) { if(n%2 == 0){
                    return (k + prev) / 2.0;
                }
                else return k;}
            prev = k;
            k = nums2[l2];
            l2++;
            t++;
        }
        if(n%2 == 0) return (k + prev) /2.0;
        return k;




        
    }
}

File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.py
class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        if len(nums1) > len(nums2) :
            nums1 , nums2 = nums2 , nums1
        n = len(nums1) + len(nums2)
        half = n // 2;
        l = 0
        r = len(nums1) - 1
        while True:
            m1 = (l + r) // 2 
            m2 = half - m1 - 2
            l1 = nums1[m1] if m1>=0  else float("-inf")
            r1 = nums1[m1 + 1] if m1 + 1 < len(nums1) else  float("inf")

            l2 = nums2[m2] if m2 >= 0 else  float("-inf")
            r2 = nums2[m2 + 1] if m2 + 1 < len(nums2) else  float("inf")

            if l1 <= r2 and l2 <= r1:
                if n%2 == 0:
                    return (max(l1,l2) + min(r1,r2))/2.0
                else:
                    return min(r1,r2)
            elif l1 > r2:
                r = m1-1
            else :
                l = m1 + 1
        return -1

        